### PR TITLE
Fix async image load

### PR DESCRIPTION
**Pull request by kroymann**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [X] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [ ] I have provided test coverage for my change (where applicable)

### Description
This addresses the issue outlined in #1997 where `Image.LoadAsync(stream)` would incorrectly use synchronous IO if the stream is seekable.  The fix is simply to remove the optimization that avoids prebuffering the stream into an in-memory buffer if the stream is seekable, and instead _always_ prebuffer the stream.  This ensures that async IO is always used to read the stream.

### CODE REVIEW

File: src/ImageSharp/Formats/IImageDecoder.cs
Comment (antonfirsov) [inline src/ImageSharp/Formats/IImageDecoder.cs:25]: Since no users are expected to call this method directly, we should avoid making the `cancellationToken` argument optional.

Comment (kroymann) [reply src/ImageSharp/Formats/IImageDecoder.cs:25]: Really?  Is that a stylistic choice within the ImageSharp codebase not to have optional arguments?  Given that these are methods on a public interface, I would have assumed that there might be people using these APIs directly, so providing a default for the `cancellationToken` seemed like a good way to minimize the breaking nature of the change.

Comment (antonfirsov) [reply src/ImageSharp/Formats/IImageDecoder.cs:24]: Yeah we tend to avoid optional arguments, because there is higher chance for making binary breaking changes. (It can be a source of endless arguments whether this is right or not, but it's not the point here and `CancellationToken` is actually an exception for this rule.)

What's important in this context is that `IImageDecoder` is not meant to be consumed, but to *be implemented*. With this in mind, avoiding `cancellationToken = default` makes things simpler.

File: src/ImageSharp/Image.FromStream.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.FromStream.cs:779]: So here we're resetting the input stream position before we attempt to decode because [`Configuration.ReadOrigin`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Configuration.html#SixLabors_ImageSharp_Configuration_ReadOrigin) has been set. We need to ensure that we still respect that configuration option.

Comment (kroymann) [reply src/ImageSharp/Image.FromStream.cs:779]: Ahhh now I get what you were saying.  Somehow I completely missed that bit.  I'll add that back in.

**Comment by JimBobSquarePants** [comment]
Everything looks good so far from an implementation perspective. We'll need to figure out a strategy now to compare the new Async performance with Main.

**Comment by JimBobSquarePants** [comment]
@antonfirsov I just pushed a change to `ChunkedMemoryStream` that allows the size of the pooled buffer chunks to scale with the total allocation.

Allocation blocks match this scale.

```
LargeChunks = Min(4Mb, MemoryAllocator.BufferCapacityInBytes())
LargeChunkThreshold = Min(1Mb, LargeChunk / 4)
SmallChunks = Min(128Kb, LargeChunks / 32)
```

Once the total allocation within the stream hits our threshold we immediately switch to allocating larger chunk sizes.

I chose the the the values for the following reasons

- Small @ 128Kb because that's the default chunk size `RecycledMemoryStream` uses. 
- Threshold @ 1Mb because that's the threshold of the ArrayPool. I was tempted to go for 512Kb here.
- Large @ 4Mb because that matches the best performing pool size for our current allocator.

I'm not precious about any of these values to if you feel like they should change then please suggest alternatives.

**Review by JimBobSquarePants — APPROVED** [review]
I'm approving this but want a second opinion before merging.

**Comment by antonfirsov** [comment]
I want to take a look next week. Quite distracted from everything programming-related thanks to the ongoimg Fourth Reich experiment on my continent.

**Comment by JimBobSquarePants** [comment]
Of course mate. Absolute horror show over there!

**Comment by antonfirsov** [comment]
@kroymann can you please give me permission to push to your branch? (Thought it is always allowed for contributors for PR branches.)
I want to push my load tests:
https://github.com/antonfirsov/ImageSharp/commit/63e897e973a6f6fbe58e5b0edcc070901281c60c

**Comment by antonfirsov** [comment]
I did before / after runs of asnchronous load testing:
- It looks like we can nicely see the result of thread pool starvation on `main` branch: The synchronous parallel run finishes in about 6.3 seconds on my 10 core machine, while the async version needs ~9.2 seconds.
- The PR seems to fix this nicely, bringing down async execution time close to the synchronous time
- Unfortunately, there is a significant *~25% memory regression*

#### Before
image

#### After
image

We should try to address it. Will take quick look at the code changes now.

**Review by antonfirsov — APPROVED** [review]
Ok, I have instant idea to improve this.

If you guys both agree that 25% extra memory is an acceptable price for correct async, I'm fine merging this. We can open an issue to track the sync/async memory gap.

Just please remember to merge my load test addition!

**Comment by antonfirsov** [comment]
Btw @kroymann thanks for discovering the problem, and pushing through the fix, really great job!

**Comment by JimBobSquarePants** [comment]
@antonfirsov Your permissions issues were likely due to Git LFS issues with GitHub. 

There's a workaround that should allow you to push to the fork.
```
rm .git/hooks/pre-push
```

**Comment by kroymann** [comment]
@antonfirsov I think @JimBobSquarePants is right about the github permissions, but I went ahead and cherry-picked your commit just to expedite things.  Thank you for adding those tests!

**Comment by antonfirsov** [comment]
> I'm assuming the memory overheads come from my buffer allocation? Would a drop to 2MB segments be useful?

2MB logical allocations lead to 4MB actual allocations since buffers are uniform.

Actually, there might be a way to create a smarter buffer growing strategy.

This is how subsequent chunks sizes are allocated currently if I get it right.
```
128K -> 128K -> 128K -> 128K -> 128K -> 128K -> 128K -> 128K ->
4MB -> 4MB -> 4MB  -> ...
```
This stressing the 128K ArrayPool unnecessarily, and often leads to a wasteful allocation in the end.

Instead, I would try to do something like:

```
128K -> 128K -> 128K -> 128K ->
256K -> 256K ->  256K -> 256K ->
512K -> 512K ->  512K -> 512K ->
1MB -> 1MB -> 1MB -> 1MB -> 
4MB -> 4MB -> 4MB  -> ...
``` 

This would keep most of the buffers rented from `ArrayPool.Shared` without stressing individual buckets too much and avoid wasting space. The unmanaged pool wouldn't be touched for images smaller than 7 MB.

**Comment by JimBobSquarePants** [comment]
Yeah, that's how it's allocated currently. Shouldn't be too hard to update it to follow your pattern.

**Comment by antonfirsov** [comment]
Here is a closed formula to get the i-th chunk size according to my logic, assuming you increment `i` each time you allocate a chunk:

```C#
private static int GetChunkSize(int i)
{
    const int _128K = 1 << 17;
    const int _4M = 1 << 22;
    return i < 16 ? _128K * (1 << (i / 4)) : _4M;
}
```

Tomorrow is the last day I have access to my benchmark machine to redo the benchmarks, I will leave for two weeks afterwards.

**Comment by JimBobSquarePants** [comment]
It's always tomorrow in Aus!

That's the changes pushed. I added a check to ensure we always use the min of the allocator capacity or the calculated size, in case we change things in the future or someone implements their own allocator.

**Comment by antonfirsov** [comment]
Looks like there is a slight improvement. The total memory gap seems to be of 20%.

image

File: src/ImageSharp/IO/ChunkedMemoryStream.cs
Comment (antonfirsov) [inline src/ImageSharp/IO/ChunkedMemoryStream.cs:550]: A message to our future selves.
```suggestion
            // Increment chunks sizes with moderate speed, but without using too many buffers from the same ArrayPool bucket of the default MemoryAllocator.
            //https://github.com/SixLabors/ImageSharp/pull/2006#issuecomment-1066244720
#pragma warning disable IDE1006 // Naming Styles
            const int _128K = 1 << 17;
```

**Review by antonfirsov — APPROVED** [review]
LGTM, thanks for the patience!

**Pull request merged**
